File: Java/Programmers/Kit/Hash/Cloths.java
/** 경우의 수 문제
* 해결책 1) 해시를 통ㅎ해 각 type 별로 가짓수를 계산하고 총 경우의 수를 산출함
*/

// key : 옷의 종류, Value : 옷 종류의 가짓수(count)

import java.util.HashMap;
import java.util.Iterator;

class Solution {
    public int solution(String[][] clothes) {
        // 1. 옷을 종류별로 구분하기
        HashMap<String, Integer> map = new HashMap<>();
        for (String[] clothe : clothes) {
            String type = clothe[1];
            map.put(type, map.getOrDefault(type, 0) + 1);
        }

        // 2. 입지 않는 경우를 추가하여 모든 조합 계산하기
        Iterator<Integer> it = map.values().iterator();
        int answer = 1;

        while(it.hasNext())
            answer *= it.next().intValue() + 1;

        // 3. 아무종류의 옷도 입지 않는 경우 제외하기
        return answer - 1;
    }
}

File: Python/Programmers/Kit/Hash/Cloths.py
# 알고리즘 고득점 KIT _ 해시
# 의상

#   경우의 수 문제
#   해결책 1) 해시를 통해 각 type 별로 가짓수를 계산하고 총 경우의 수를 산출함
#   key : 옷의 종류, Value : 옷 종류의 가짓수(count)

def solution(clothes):
    answer = 0
    hashmap = {}

    for i in range(len(clothes)):
        keyx = clothes[i][1]
        # 해쉬맵에 키값 기존재 안하는 경우
        if keyx not in hashmap:
            hashmap[keyx] = 1
        else:                       # 기존재 하는 경우
            temp = hashmap[keyx]    # 벨류값
            hashmap[keyx] = temp +1
    # print(hashmap)#	{'headgear': 2, 'eyewear': 1}

    ## 해쉬맵 경우의수  3*2 -1 (아무것도 입지않는경우 1개 빼주기)
    result = 1
    for j in hashmap:
        result *= hashmap[j]+1
        # print(result, hashmap[j])

    answer = result-1

    return answer

# 2024.02.08 : 16m